rpcs3/util/expr.hpp

```diff
@@ -0,0 +1,193 @@
+#pragma once
+
+#include "util/types.hpp"
+
+template <usz Size>
+struct utf32_expr
+{
+	static constexpr usz size = Size;
+
+	static_assert(Size > 0);
+
+	char32_t data[Size]{};
+
+	template <usz N>
+	constexpr utf32_expr(const char32_t(&str)[N])
+	{
+		for (usz i = 0; i < Size; i++)
+		{
+			data[i] = str[i];
+		}
+	}
+
+	template <typename CharT, usz N> requires (sizeof(CharT) == 1)
+	constexpr utf32_expr(const CharT(&str)[N])
+	{
+		for (usz i = 0, j = 0; i < Size && j < Size; j++)
+		{
+			const uchar ch = static_cast<uchar>(str[j]);
+
+			if (ch >= 0 && ch <= 0x7f)
+			{
+				// Write single ASCII character
+				data[i++] = ch;
+			}
+			else if ((ch & 0xc0) == 0x80)
+			{
+				// Append UTF-8 encoded tail
+				data[i] = (data[i] << 6) | ch;
+			}
+			else
+			{
+				// Write UTF-8 head
+				const uint sh = std::countl_one(ch);
+				data[i++] = ((ch << sh) & 0xff) >> sh;
+			}
+		}
+	}
+};
+
+template <usz Size>
+utf32_expr(const char32_t(&a)[Size]) -> utf32_expr<Size>;
+
+template <typename CharT, usz Size> requires (sizeof(CharT) == 1)
+utf32_expr(const CharT(&a)[Size]) -> utf32_expr<Size>;
+
+template <char32_t Cmd, utf32_expr S, usz I = 0>
```

Comment:
If `I` has default value of 1 then the first command will always get executed twice. Once `utf32_cmd<S.data[0], S, 0>` then again `utf32_cmd<S.data[0], S, 1>`